File: swap_flutter/lib/core/utils/error_handler.dart
import 'package:dio/dio.dart';
import 'package:logger/logger.dart';

/// Global Error Handler
/// 
/// Centralized error handling system for consistent error processing
/// and user-friendly error messages throughout the app.
class ErrorHandler {
  static const _networkErrorMessages = {
    'connection_timeout': 'Connection timeout. Please check your internet connection.',
    'send_timeout': 'Request timeout. Please try again.',
    'receive_timeout': 'Server response timeout. Please try again.',
    'no_internet': 'No internet connection. Please check your connection.',
    'server_error': 'Server error occurred. Please try again later.',
    'unauthorized': 'Session expired. Please log in again.',
    'forbidden': 'Access denied. You don\'t have permission to perform this action.',
    'not_found': 'The requested resource was not found.',
    'too_many_requests': 'Too many requests. Please wait a moment and try again.',
    'unknown': 'An unexpected error occurred. Please try again.',
  };

  static const _validationErrorMessages = {
    'email_invalid': 'Please enter a valid email address.',
    'password_weak': 'Password must be at least 8 characters long.',
    'phone_invalid': 'Please enter a valid phone number.',
    'amount_invalid': 'Please enter a valid amount.',
    'field_required': 'This field is required.',
  };

  /// Handle DioException and return user-friendly error message
  static AppError handleDioError(DioException error) {
    switch (error.type) {
      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        return AppError(
          type: ErrorType.network,
          code: error.type.name,
          message: _networkErrorMessages[error.type.name] ?? _networkErrorMessages['unknown']!,
          originalError: error,
        );

      case DioExceptionType.badResponse:
        return _handleHttpError(error);

      case DioExceptionType.cancel:
        return AppError(
          type: ErrorType.cancelled,
          code: 'request_cancelled',
          message: 'Request was cancelled',
          originalError: error,
        );

      case DioExceptionType.connectionError:
        return AppError(
          type: ErrorType.network,
          code: 'no_internet',
          message: _networkErrorMessages['no_internet']!,
          originalError: error,
        );

      case DioExceptionType.badCertificate:
        return AppError(
          type: ErrorType.security,
          code: 'bad_certificate',
          message: 'Security certificate error. Please check your connection.',
          originalError: error,
        );

      case DioExceptionType.unknown:
      default:
        return AppError(
          type: ErrorType.unknown,
          code: 'unknown',
          message: _networkErrorMessages['unknown']!,
          originalError: error,
        );
    }
  }

  /// Handle HTTP status code errors
  static AppError _handleHttpError(DioException error) {
    final statusCode = error.response?.statusCode;
    final responseData = error.response?.data;

    switch (statusCode) {
      case 400:
        return AppError(
          type: ErrorType.validation,
          code: 'bad_request',
          message: _extractErrorMessage(responseData) ?? 'Invalid request. Please check your input.',
          originalError: error,
          details: responseData,
        );

      case 401:
        return AppError(
          type: ErrorType.authentication,
          code: 'unauthorized',
          message: _networkErrorMessages['unauthorized']!,
          originalError: error,
        );

      case 403:
        return AppError(
          type: ErrorType.authorization,
          code: 'forbidden',
          message: _networkErrorMessages['forbidden']!,
          originalError: error,
        );

      case 404:
        return AppError(
          type: ErrorType.notFound,
          code: 'not_found',
          message: _networkErrorMessages['not_found']!,
          originalError: error,
        );

      case 422:
        return AppError(
          type: ErrorType.validation,
          code: 'validation_error',
          message: _extractErrorMessage(responseData) ?? 'Validation failed. Please check your input.',
          originalError: error,
          details: responseData,
        );

      case 429:
        return AppError(
          type: ErrorType.rateLimit,
          code: 'too_many_requests',
          message: _networkErrorMessages['too_many_requests']!,
          originalError: error,
        );

      case 500:
      case 502:
      case 503:
      case 504:
        return AppError(
          type: ErrorType.server,
          code: 'server_error',
          message: _networkErrorMessages['server_error']!,
          originalError: error,
        );

      default:
        return AppError(
          type: ErrorType.unknown,
          code: 'http_error',
          message: _extractErrorMessage(responseData) ?? _networkErrorMessages['unknown']!,
          originalError: error,
        );
    }
  }

  /// Extract error message from response data
  static String? _extractErrorMessage(dynamic responseData) {
    if (responseData == null) return null;
    
    if (responseData is Map<String, dynamic>) {
      // Common API error message fields
      final messageFields = ['message', 'error', 'detail', 'msg'];
      
      for (final field in messageFields) {
        if (responseData.containsKey(field) && responseData[field] is String) {
          return responseData[field];
        }
      }

      // Handle validation errors
      if (responseData.containsKey('errors') && responseData['errors'] is Map) {
        final errors = responseData['errors'] as Map<String, dynamic>;
        final firstError = errors.values.first;
        if (firstError is List && firstError.isNotEmpty) {
          return firstError.first.toString();
        }
      }
    }

    return null;
  }

  /// Handle generic exceptions
  static AppError handleGenericError(Object error, StackTrace? stackTrace) {
    if (error is DioException) {
      return handleDioError(error);
    }

    if (error is AppError) {
      return error;
    }

    return AppError(
      type: ErrorType.unknown,
      code: 'generic_error',
      message: error.toString(),
      originalError: error,
      stackTrace: stackTrace,
    );
  }

  /// Get validation error message for a specific field
  static String getValidationErrorMessage(String errorCode) {
    return _validationErrorMessages[errorCode] ?? 
           _validationErrorMessages['field_required']!;
  }

  /// Log error with appropriate level
  static void logError(AppError error, Logger logger) {
    switch (error.type) {
      case ErrorType.network:
      case ErrorType.server:
        logger.w('${error.type.name.toUpperCase()}: ${error.message}', error: error.originalError);
        break;
      
      case ErrorType.authentication:
      case ErrorType.authorization:
        logger.i('${error.type.name.toUpperCase()}: ${error.message}');
        break;
      
      case ErrorType.validation:
        logger.d('${error.type.name.toUpperCase()}: ${error.message}');
        break;
      
      case ErrorType.security:
      case ErrorType.unknown:
        logger.e('${error.type.name.toUpperCase()}: ${error.message}', 
                error: error.originalError, 
                stackTrace: error.stackTrace);
        break;
      
      default:
        logger.w('${error.type.name.toUpperCase()}: ${error.message}', error: error.originalError);
    }
  }
}

/// Application Error Class
class AppError implements Exception {
  final ErrorType type;
  final String code;
  final String message;
  final Object? originalError;
  final StackTrace? stackTrace;
  final Map<String, dynamic>? details;

  const AppError({
    required this.type,
    required this.code,
    required this.message,
    this.originalError,
    this.stackTrace,
    this.details,
  });

  @override
  String toString() {
    return 'AppError(type: $type, code: $code, message: $message)';
  }

  /// Create a copy with updated fields
  AppError copyWith({
    ErrorType? type,
    String? code,
    String? message,
    Object? originalError,
    StackTrace? stackTrace,
    Map<String, dynamic>? details,
  }) {
    return AppError(
      type: type ?? this.type,
      code: code ?? this.code,
      message: message ?? this.message,
      originalError: originalError ?? this.originalError,
      stackTrace: stackTrace ?? this.stackTrace,
      details: details ?? this.details,
    );
  }
}

/// Error Types Enumeration
enum ErrorType {
  network,
  server,
  authentication,
  authorization,
  validation,
  notFound,
  rateLimit,
  security,
  cancelled,
  unknown,
}

/// Error Result for API responses
sealed class Result<T> {
  const Result();
}

class Success<T> extends Result<T> {
  final T data;
  
  const Success(this.data);
}

class Failure<T> extends Result<T> {
  final AppError error;
  
  const Failure(this.error);
}

File: swap-expo/_app/utils/errorHandler.ts
/**
 * Standardized Error Handling Utility
 * 
 * Provides consistent error handling patterns across the application.
 * Integrates with logger and TanStack Query for unified error management.
 */

import logger from './logger';
import { HttpError } from '../types/api.types';

// Standard error categories for better organization
export enum ErrorCategory {
  AUTH = 'auth',
  API = 'api', 
  NETWORK = 'network',
  STORAGE = 'storage',
  VALIDATION = 'validation',
  PERMISSION = 'permission',
  UNKNOWN = 'unknown'
}

// Error severity levels
export enum ErrorSeverity {
  LOW = 'low',       // Non-critical, can continue
  MEDIUM = 'medium', // Important but recoverable  
  HIGH = 'high',     // Critical, affects core functionality
  CRITICAL = 'critical' // App-breaking, requires immediate attention
}

// Standardized error interface
export interface AppError extends Error {
  category: ErrorCategory;
  severity: ErrorSeverity;
  code?: string;
  statusCode?: number;
  metadata?: Record<string, any>;
  userMessage?: string; // User-friendly message
  originalError?: Error;
}

/**
 * Create a standardized AppError
 */
export const createAppError = (
  message: string,
  category: ErrorCategory = ErrorCategory.UNKNOWN,
  severity: ErrorSeverity = ErrorSeverity.MEDIUM,
  options?: {
    code?: string;
    statusCode?: number;
    metadata?: Record<string, any>;
    userMessage?: string;
    originalError?: Error;
  }
): AppError => {
  const error = new Error(message) as AppError;
  error.category = category;
  error.severity = severity;
  error.code = options?.code;
  error.statusCode = options?.statusCode;
  error.metadata = options?.metadata;
  error.userMessage = options?.userMessage || message;
  error.originalError = options?.originalError;
  return error;
};

/**
 * Convert unknown error to standardized AppError
 */
export const normalizeError = (
  error: unknown,
  category: ErrorCategory = ErrorCategory.UNKNOWN,
  context?: string
): AppError => {
  // Already an AppError
  if (isAppError(error)) {
    return error;
  }

  // Axios/HTTP error
  if (isHttpError(error)) {
    return createAppError(
      error.message || 'HTTP request failed',
      ErrorCategory.API,
      getHttpErrorSeverity(error.status || 500),
      {
        statusCode: error.status,
        code: error.code,
        metadata: { url: error.config?.url, context },
        userMessage: getHttpErrorMessage(error.status || 500),
        originalError: error
      }
    );
  }

  // Standard Error object
  if (error instanceof Error) {
    return createAppError(
      error.message,
      category,
      ErrorSeverity.MEDIUM,
      {
        metadata: { context },
        originalError: error
      }
    );
  }

  // String error
  if (typeof error === 'string') {
    return createAppError(
      error,
      category,
      ErrorSeverity.MEDIUM,
      {
        metadata: { context }
      }
    );
  }

  // Unknown error type
  return createAppError(
    'An unknown error occurred',
    category,
    ErrorSeverity.MEDIUM,
    {
      metadata: { originalError: error, context }
    }
  );
};

/**
 * Check if error is an AppError
 */
export const isAppError = (error: unknown): error is AppError => {
  return error instanceof Error && 'category' in error && 'severity' in error;
};

/**
 * Check if error is an HTTP/Axios error
 */
export const isHttpError = (error: unknown): error is HttpError => {
  return (
    error !== null &&
    typeof error === 'object' &&
    ('status' in error || 'response' in error || 'code' in error)
  );
};

/**
 * Get error severity based on HTTP status code
 */
const getHttpErrorSeverity = (statusCode: number): ErrorSeverity => {
  if (statusCode >= 500) return ErrorSeverity.HIGH;
  if (statusCode >= 400) return ErrorSeverity.MEDIUM;
  return ErrorSeverity.LOW;
};

/**
 * Get user-friendly message for HTTP errors
 */
const getHttpErrorMessage = (statusCode: number): string => {
  switch (statusCode) {
    case 400:
      return 'Invalid request. Please check your input and try again.';
    case 401:
      return 'Authentication required. Please sign in and try again.';
    case 403:
      return 'Access denied. You don\'t have permission for this action.';
    case 404:
      return 'The requested resource was not found.';
    case 408:
      return 'Request timeout. Please check your connection and try again.';
    case 429:
      return 'Too many requests. Please wait a moment and try again.';
    case 500:
      return 'Server error. Please try again later.';
    case 502:
    case 503:
    case 504:
      return 'Service temporarily unavailable. Please try again later.';
    default:
      return 'An error occurred. Please try again.';
  }
};

/**
 * Standardized error logging
 */
export const logError = (
  error: AppError | unknown,
  context?: string,
  metadata?: Record<string, any>
): void => {
  const appError = isAppError(error) ? error : normalizeError(error, ErrorCategory.UNKNOWN, context);
  
  const logData = {
    category: appError.category,
    severity: appError.severity,
    code: appError.code,
    statusCode: appError.statusCode,
    context,
    ...appError.metadata,
    ...metadata
  };

  // Log based on severity
  switch (appError.severity) {
    case ErrorSeverity.CRITICAL:
    case ErrorSeverity.HIGH:
      logger.error(appError.message, appError.originalError, appError.category, logData);
      break;
    case ErrorSeverity.MEDIUM:
      logger.warn(appError.message, appError.category, logData);
      break;
    case ErrorSeverity.LOW:
      logger.debug(appError.message, appError.category, logData);
      break;
  }
};

/**
 * Standard retry logic for TanStack Query
 */
export const createRetryFunction = (maxRetries: number = 2) => {
  return (failureCount: number, error: unknown) => {
    const appError = normalizeError(error, ErrorCategory.API);
    
    // Don't retry client errors (4xx)
    if (appError.statusCode && appError.statusCode >= 400 && appError.statusCode < 500) {
      return false;
    }
    
    // Don't retry critical errors
    if (appError.severity === ErrorSeverity.CRITICAL) {
      return false;
    }
    
    return failureCount < maxRetries;
  };
};

/**
 * Extract user-friendly error message for UI display
 */
export const getUserErrorMessage = (error: unknown): string => {
  if (isAppError(error)) {
    return error.userMessage || error.message;
  }
  
  const appError = normalizeError(error);
  return appError.userMessage || 'An unexpected error occurred';
};

/**
 * Safe async function wrapper with standardized error handling
 */
export const safeAsync = <T>(
  fn: () => Promise<T>,
  category: ErrorCategory = ErrorCategory.UNKNOWN,
  context?: string
): Promise<T> => {
  return fn().catch((error) => {
    const appError = normalizeError(error, category, context);
    logError(appError, context);
    throw appError;
  });
};

/**
 * Storage operation error handler
 */
export const handleStorageError = (error: unknown, operation: string): void => {
  const appError = normalizeError(error, ErrorCategory.STORAGE, `Storage ${operation}`);
  logError(appError);
  
  // For storage errors, we usually want to continue execution
  // but log the error for debugging
};

/**
 * Auth operation error handler
 */
export const handleAuthError = (error: unknown, operation: string): AppError => {
  const appError = normalizeError(error, ErrorCategory.AUTH, `Auth ${operation}`);
  logError(appError);
  return appError;
};

/**
 * Network operation error handler  
 */
export const handleNetworkError = (error: unknown, operation: string): AppError => {
  const appError = normalizeError(error, ErrorCategory.NETWORK, `Network ${operation}`);
  logError(appError);
  return appError;
};

export default {
  createAppError,
  normalizeError,
  logError,
  createRetryFunction,
  getUserErrorMessage,
  safeAsync,
  handleStorageError,
  handleAuthError,
  handleNetworkError,
  ErrorCategory,
  ErrorSeverity
};